envconfig.go

```diff
@@ -32,75 +54,227 @@ type Decoder interface {
 	Decode(value string) error
 }
 
+// ProcessFunc vistor function definition while traversing the list of options
+type ProcessFunc func(field reflect.Value, tof reflect.StructField, fieldName string, key string, alt_key string, def string, required string) error
+
 func (e *ParseError) Error() string {
 	return fmt.Sprintf("envconfig.Process: assigning %[1]s to %[2]s: converting '%[3]s' to type %[4]s", e.KeyName, e.FieldName, e.Value, e.TypeName)
 }
 
-// Process populates the specified struct based on environment variables
-func Process(prefix string, spec interface{}) error {
+// Visit vists all the configuration options, calling the process function on each
+func Visit(prefix string, spec interface{}, process ProcessFunc) error {
```

Comment:
But at that point someone is already writing their code in a separate package. Why should envconfig be responsible for providing a general visitor pattern implementation?

envconfig's job is reading environment variables into a struct. This represents a pretty big departure from that mission, a big broadening of the scope.